### ISSUE TITLE
getBranches does not use git auth url

### ISSUE THREAD

**Issue by dotfold** [semantic-release/semantic-release#1170]
## Current behavior

I'm running semantic-release in a `node:11` docker container, with `GL_URL` and `GL_TOKEN` values provided.

During the setup phase after [`verify(context)`](https://github.com/semantic-release/semantic-release/blob/beta/index.js#L54), [`getBranches`](https://github.com/semantic-release/semantic-release/blob/beta/index.js#L54) is then called which eventually results in the invocation of the git command `git ls-remote --heads origin`. This is made without any auth and results in an authentication error from the remote.

I've followed the code path through, and the command is invoked in [`lib/branches/expand.js`](https://github.com/semantic-release/semantic-release/blob/beta/lib/branches/expand.js#L6)

Should `getGitAuthUrl` be run first, and then the `options.repositoryUrl` be passed through to `getBranches` and subsequently also the call to `fetch` in [`lib/branches/index`](https://github.com/semantic-release/semantic-release/blob/beta/lib/branches/index.js#L20)?

I've added that change in locally and `getBranches` works when using the `gitAuthUrl` instead of `origin`.

I'd be happy to submit a PR if this approach is valid.

## Expected behavior

There should not be an authentication error when a token is supplied.

## Environment

- **semantic-release** version: 16.0.0-beta.18
- CI environment: docker
- Plugins used: 
- **semantic-release** configuration: 
```
---
branches:
  - master
  - name: next
    prerelease: 'rc'
  - name: beta
    prerelease: 'beta'
plugins:
  - '@semantic-release/commit-analyzer'
  - '@semantic-release/release-notes-generator'
  - '@semantic-release/gitlab'
success: false
fail: false
```
- CI logs: 

Note: this is running in a `node:11` docker container.

```
root@94e0a71c0f6d:/src# GL_URL=<REDACTED> GL_TOKEN=<REDACTED> npm run release:dry

> beta-channel-example@1.0.0 release:dry /src
> semantic-release --debug --dry-run

[10:38:19 PM] [semantic-release] › ℹ  Running semantic-release version 16.0.0-beta.18
  semantic-release:config load config from: /src/.releaserc.yaml +0ms
  semantic-release:config options values: { branches:
  semantic-release:config    [ 'master',
  semantic-release:config      { name: 'next', prerelease: 'rc' },
  semantic-release:config      { name: 'beta', prerelease: 'beta' } ],
  semantic-release:config   repositoryUrl:
  semantic-release:config    '<REDACTED>',
  semantic-release:config   tagFormat: 'v${version}',
  semantic-release:config   plugins:
  semantic-release:config    [ '@semantic-release/commit-analyzer',
  semantic-release:config      '@semantic-release/release-notes-generator',
  semantic-release:config      '@semantic-release/gitlab' ],
  semantic-release:config   success: false,
  semantic-release:config   fail: false,
  semantic-release:config   _: [],
  semantic-release:config   debug: true,
  semantic-release:config   'dry-run': true,
  semantic-release:config   d: true,
  semantic-release:config   dryRun: true,
  semantic-release:config   '$0': 'node_modules/.bin/semantic-release' } +15ms
  semantic-release:plugins options for @semantic-release/gitlab/verifyConditions: {} +0ms
[10:38:20 PM] [semantic-release] › ✔  Loaded plugin "verifyConditions" from "@semantic-release/gitlab"
  semantic-release:plugins options for @semantic-release/commit-analyzer/analyzeCommits: {} +7ms
[10:38:20 PM] [semantic-release] › ✔  Loaded plugin "analyzeCommits" from "@semantic-release/commit-analyzer"
  semantic-release:plugins options for @semantic-release/release-notes-generator/generateNotes: {} +2ms
[10:38:20 PM] [semantic-release] › ✔  Loaded plugin "generateNotes" from "@semantic-release/release-notes-generator"
  semantic-release:plugins options for @semantic-release/gitlab/publish: {} +1ms
[10:38:20 PM] [semantic-release] › ✔  Loaded plugin "publish" from "@semantic-release/gitlab"
[10:38:22 PM] [semantic-release] › ✖  An error occurred while running semantic-release: { Error: Command failed: git ls-remote --heads origin
remote: HTTP Basic: Access denied
fatal: Authentication failed for '<REDACTED>'


    at makeError (/src/node_modules/execa/index.js:174:9)
    at Promise.all.then.arr (/src/node_modules/execa/index.js:278:16)
    at processTicksAndRejections (internal/process/next_tick.js:81:5)
  code: 128,
  stdout: '',
  stderr:
   "remote: HTTP Basic: Access denied\nfatal: Authentication failed for '<REDACTED>'\n",
  failed: true,
  signal: null,
  cmd: 'git ls-remote --heads origin',
  timedOut: false,
  killed: false }
{ Error: Command failed: git ls-remote --heads origin
remote: HTTP Basic: Access denied
fatal: Authentication failed for '<REDACTED>'


    at makeError (/src/node_modules/execa/index.js:174:9)
    at Promise.all.then.arr (/src/node_modules/execa/index.js:278:16)
    at processTicksAndRejections (internal/process/next_tick.js:81:5)
  code: 128,
  stdout: '',
  stderr:
   "remote: HTTP Basic: Access denied\nfatal: Authentication failed for '<REDACTED>'\n",
  failed: true,
  signal: null,
  cmd: 'git ls-remote --heads origin',
  timedOut: false,
  killed: false }npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! beta-channel-example@1.0.0 release:dry: `semantic-release --debug --dry-run`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the beta-channel-example@1.0.0 release:dry script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2019-05-01T22_38_22_432Z-debug.log
```

**Comment by jmodjeski75** [comment]
I am encountering the same issue but with a GitHub repo instead. Do you have a gist of the purposed change?

**Comment by gr2m** [comment]
If you could start a PR that’d be great, we can continue the discussion there :)

**Comment by jmodjeski75** [comment]
Currently working on resolving all the places where origin is used instead of the correct auth url.

**Comment by alexan** [comment]
did anybody figure out a workaround for this problem?

**Comment by jmodjeski75** [comment]
@alexan a PR is open #1186 you are welcome to leverage my fork but it is as is until it is merged here.

**Comment by gr2m** [comment]
Alexander, please let us know if @jmodjeski75’s fork fixed the issue for you

**Comment by alexan** [comment]
@gr2m I tested the fork with devDependencies`"semantic-release": "https://github.com/jmodjeski75/semantic-release"` and did not get any error from `git ls-remote --heads origin` and successfully release a test package which previously failed

**Comment by pvdlg** [comment]
Fixed by #1186 in beta
